cirq-core/cirq/ops/post_selection_gate.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, Iterable, Sequence, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality()
+class PostSelectionGate(raw_types.Gate):
+    r"""A gate simulating post-selection from a given subspace or set of subspaces.
+
+    This gate is only for simulation, and cannot be run on a real device as it violates the laws of
+    quantum mechanics.
+    """
+
+    def __init__(self, qid_shape: Sequence[int], subspaces: Iterable[Sequence[int]]):
+        r"""Creates a gate simulating post-selection.
+
+        Args:
+            qid_shape: The shape of qubits this gate applies to.
+            subspaces: The post-selection subspace, provided as a set of index tuples. For example,
+                |01><01| + |22><22| on a set of 3D qudits would be passed in as [(0, 1), (2, 2)].
+                The waveform will be projected onto the subspace and renormalized. If the
+                projection onto the subspace is zero, a ValueError will be raised.
+        """
+        self._subspaces = tuple(sorted(set(tuple(s) for s in subspaces)))
+        self._qid_shape = tuple(qid_shape)
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _has_unitary_(self) -> bool:
+        return True
```

Comment:
This is the dangerous part of this PR - several parts of Cirq rely on this to determine whether a gate is unitary. The most immediately relevant case I can think of is in the [unitary prefix part of simulation](https://github.com/quantumlib/Cirq/blob/6a97ccaa7e24ddf33656786a5b1cfb50cf609dee/cirq-core/cirq/sim/simulator_base.py#L171).

Reply:
Yeah, this lies in a weird space beyond `_unitary_` / `_mixture_` / `_kraus_`, where the behavior is non-unitary but the result is a single state. It's plausible that we could accomplish this with some magic on top of channels, but an RFC is definitely justified.